src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -413,62 +779,101 @@ def validate(self, node, options=None):
+                        f"it accesses variable '{sym.name}' from its "
                         f"parent container.")
 
         # Check that the shape of any dummy array arguments are the same as
         # those at the call site.
         visitor = FortranWriter()
         for dummy_arg, actual_arg in zip(routine_table.argument_list,
                                          node.children):
-            # TODO #1799 this really needs the `datatype` method that is
-            # planned for all reference nodes.
-            if isinstance(actual_arg, ArrayReference):
-                rank = 0
-                for idx, expr in enumerate(actual_arg.indices):
-                    if isinstance(expr, Range):
-                        rank += 1
-                        # TODO #924 add a method to Range that returns the
-                        # PSyIR expression for the number of elements it
-                        # contains. This can then be compared with the shape of
-                        # the dummy argument.
-                        if not actual_arg.is_full_range(idx):
-                            # It's OK to use the loop variable in the lambda
-                            # definition because if we get to this point then
-                            # we're going to quit the loop.
-                            # pylint: disable=cell-var-from-loop
-                            raise TransformationError(LazyString(
-                                lambda: f"Cannot inline routine "
-                                f"'{routine.name}' because argument "
-                                f"'{visitor(actual_arg)}' is "
-                                f"an array subsection (TODO #924)."))
-                if isinstance(dummy_arg.datatype, ScalarType):
-                    dummy_rank = 0
-                else:
-                    dummy_rank = len(dummy_arg.datatype.shape)
-                if rank != dummy_rank:
-                    # It's OK to use the loop variable in the lambda definition
-                    # because if we get to this point then we're going to quit
-                    # the loop.
-                    # pylint: disable=cell-var-from-loop
-                    raise TransformationError(LazyString(
+            dummy_rank = 0
+            actual_rank = 0
+            if not isinstance(actual_arg, Reference):
+                # TODO #1799 this really needs the `datatype` method to be
+                # extended to support all nodes. For now we have to skip
+                # anything that's not a Reference.
+                continue
+
+            if (isinstance(dummy_arg.datatype, DeferredType) or
+                    isinstance(actual_arg.datatype, DeferredType)):
+                # If we haven't resolved the type of either the actual or
+                # dummy argument (e.g. because their type is imported from a
+                # module) then we can't performing type checking.
+                continue
```

Comment:
Ah yes, I was being a bit gung-ho wasn't I? My earlier refactoring has already improved/refined the second case so that it is safe. I've changed the first case to raise an exception.

Reply:
Hmm, I've been thinking about my assertion that the second case is safe. I think it *is* safe if we assume there's no array re-shaping. However, if a routine doesn't have an explicit interface (e.g. isn't in a Fortran module) then I don't think that assumption holds. I'll refine the validation checks...

Reply:
Admittedly, since we can currently only inline a Routine if it's in the same Container, I think that effectively re-shaping isn't allowed. However, this situation might change in the future so better safe than sorry.